build/protobuf.mk

```diff
@@ -0,0 +1,61 @@
+# Copyright 2015 The Cockroach Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License. See the AUTHORS file
+# for names of contributors.
+#
+# Author: Tamir Duberstein (tamird@gmail.com)
+
+# This file is evaluated in the repo's parent directory. See main.go's
+# go:generate invocation.
+
+ORG_ROOT       := .
+REPO_ROOT      := $(ORG_ROOT)/cockroach
+GITHUB_ROOT    := $(ORG_ROOT)/..
+GOGOPROTO_ROOT := $(GITHUB_ROOT)/gogo/protobuf
+
+ENGINE_ROOT := $(REPO_ROOT)/storage/engine
+
+# Ensure we have an unambiguous GOPATH
+GOPATH := $(GITHUB_ROOT)/../..
+#                        ^  ^~ GOPATH
+#                        |~ GOPATH/src
+
+GOPATH_BIN      := $(GOPATH)/bin
+PROTOC          := $(GOPATH_BIN)/protoc
+PLUGIN_SUFFIX   := gogo
+PROTOC_PLUGIN   := $(GOPATH_BIN)/protoc-gen-$(PLUGIN_SUFFIX)
+GOGOPROTO_PROTO := $(GOGOPROTO_ROOT)/gogoproto/gogo.proto
+GOGOPROTO_PATH  := $(GOGOPROTO_ROOT):$(GOGOPROTO_ROOT)/protobuf
+
+GO_PROTOS  := $(sort $(shell find $(REPO_ROOT) -not -path '*/.*' -name *.proto -type f))
+GO_SOURCES  := $(GO_PROTOS:%.proto=%.pb.go)
+
+CPP_PROTOS := $(filter %api.proto %data.proto %internal.proto %config.proto %errors.proto,$(GO_PROTOS))
+CPP_HEADERS := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.h)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.h))
+CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc)) $(subst $(GOGOPROTO_ROOT),$(ENGINE_ROOT),$(GOGOPROTO_PROTO:%.proto=%.pb.cc))
+
+.PHONY:
+protos: $(GO_SOURCES) $(CPP_HEADERS) $(CPP_SOURCES)
+
+$(GO_SOURCES): $(PROTOC) $(GO_PROTOS) $(GOGOPROTO_PROTO) $(PROTOC_PLUGIN)
+	find $(REPO_ROOT) -not -path '*/.*' -name *.pb.go | xargs rm
+	$(PROTOC) -I.:$(GOGOPROTO_PATH) --plugin=$(PROTOC_PLUGIN) --$(PLUGIN_SUFFIX)_out=$(ORG_ROOT) $(GO_PROTOS)
```

Comment:
Do you mean because it'll bundle the protos from `cockroach.storage{,.engine}` together? You could add `-maxdepth` and (maybe) `-prune` to `find`s arguments to restrict it, and have the loop over all directories which contain at least one proto file at base level, which can also be done with some `find -exec | uniq`.